crypto/fipsmodule/cpucap/internal.h

```diff
@@ -0,0 +1,173 @@
+//     ECX for CPUID where EAX = 7
+//
+// Note: the CPUID bits are pre-adjusted for the OSXSAVE bit and the YMM and XMM
+// bits in XCR0, so it is not necessary to check those.
+extern uint32_t OPENSSL_ia32cap_P[4];
+
+// See Intel manual, volume 2A, table 3-11.
+
+OPENSSL_INLINE int CRYPTO_is_FXSR_capable(void) {
+  return (OPENSSL_ia32cap_P[0] & (1 << 24)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_intel_cpu(void) {
+  // The reserved bit 30 is used to indicate an Intel CPU.
+  return (OPENSSL_ia32cap_P[0] & (1 << 30)) != 0;
+}
+
+// See Intel manual, volume 2A, table 3-10.
+
+OPENSSL_INLINE int CRYPTO_is_PCLMUL_capable(void) {
+  return (OPENSSL_ia32cap_P[1] & (1 << 1)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_SSSE3_capable(void) {
+  return (OPENSSL_ia32cap_P[1] & (1 << 9)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_SSE4_1_capable(void) {
+  return (OPENSSL_ia32cap_P[1] & (1 << 19)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_MOVBE_capable(void) {
+  return (OPENSSL_ia32cap_P[1] & (1 << 22)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_AESNI_capable(void) {
+  return (OPENSSL_ia32cap_P[1] & (1 << 25)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_AVX_capable(void) {
+  return (OPENSSL_ia32cap_P[1] & (1 << 28)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_RDRAND_capable(void) {
+  return (OPENSSL_ia32cap_P[1] & (1u << 30)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_AMD_XOP_support(void) {
+  return (OPENSSL_ia32cap_P[1] & (1 << 11)) != 0;
+}
+
+// See Intel manual, volume 2A, table 3-8.
+
+OPENSSL_INLINE int CRYPTO_is_BMI1_capable(void) {
+  return (OPENSSL_ia32cap_P[2] & (1 << 3)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_AVX2_capable(void) {
+  return (OPENSSL_ia32cap_P[2] & (1 << 5)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_BMI2_capable(void) {
+  return (OPENSSL_ia32cap_P[2] & (1 << 8)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_ADX_capable(void) {
+  return (OPENSSL_ia32cap_P[2] & (1 << 19)) != 0;
+}
+
+OPENSSL_INLINE int CRYPTO_is_SHAEXT_capable(void) {
+  return (OPENSSL_ia32cap_P[2] & (1 << 29)) != 0;
+}
+
+#endif  // OPENSSL_X86 || OPENSSL_X86_64
+
+#if defined(OPENSSL_ARM) || defined(OPENSSL_AARCH64)
+
+#if defined(OPENSSL_APPLE) && defined(OPENSSL_ARM)
+// We do not detect any features at runtime for Apple's 32-bit ARM platforms. On
+// 64-bit ARM, we detect some post-ARMv8.0 features.
+#define OPENSSL_STATIC_ARMCAP
+#endif
+
+// Normalize some older feature flags to their modern ACLE values.
+// https://developer.arm.com/architectures/system-architectures/software-standards/acle
+#if defined(__ARM_NEON__) && !defined(__ARM_NEON)
+#define __ARM_NEON 1
+#endif
+#if defined(__ARM_FEATURE_CRYPTO)
+#if !defined(__ARM_FEATURE_AES)
+#define __ARM_FEATURE_AES 1
+#endif
+#if !defined(__ARM_FEATURE_SHA2)
+#define __ARM_FEATURE_SHA2 1
+#endif
+#endif
+
+#include <openssl/arm_arch.h>
+
+extern uint32_t OPENSSL_armcap_P;
+
+// CRYPTO_is_NEON_capable returns true if the current CPU has a NEON unit. If
+// this is known statically, it is a constant inline function.
```

Comment:
Should this comment be changed? it is a
```suggestion
// CRYPTO_is_NEON_capable returns true if the current CPU has a NEON unit. If
// this is known statically, it is a constant inline function.
// Otherwise, the capability is checked at runtime by checking the corresponding bit
// in |OPENSSL_armcap_P|.
// This is also the same for |CRYPTO_is_ARMv8_AES_capable| and |CRYPTO_is_ARMv8_PMULL_capable|
// for checking for AES and PMULL instructions respectively.
```

Reply:
done